lib/src/migrator_base.dart

```diff
@@ -0,0 +1,113 @@
+// Copyright 2019 Google LLC
+//
+// Use of this source code is governed by an MIT-style
+// license that can be found in the LICENSE file or at
+// https://opensource.org/licenses/MIT.
+
+import 'dart:io';
+
+// The sass package's API is not necessarily stable. It is being imported with
+// the Sass team's explicit knowledge and approval. See
+// https://github.com/sass/dart-sass/issues/236.
+import 'package:sass/src/ast/sass.dart';
+import 'package:sass/src/syntax.dart';
+import 'package:sass/src/visitor/recursive_statement.dart';
+import 'package:sass/src/visitor/interface/expression.dart';
+
+import 'package:path/path.dart' as p;
+
+import 'migrator.dart';
+import 'patch.dart';
+import 'utils.dart';
+
+/// This base migrator parses the stylesheet at each provided entrypoint and
+/// recursively visits every statement and expression it contains.
+///
+/// Migrators based on this should add appropriate patches to [patches] in
+/// overridden methods.
+///
+/// On its own, this migrator only touches the entrypoints that are directly
+/// provided to it; it does not migrate dependencies.
+abstract class MigratorBase extends RecursiveStatementVisitor
+    with Migrator, ExpressionVisitor {
+  List<Patch> _patches = [];
+
+  /// The patches to be applied to the stylesheet being migrated.
+  ///
+  /// Subclasses that override this should also override [migrateFile].
+  List<Patch> get patches => _patches;
+
+  /// Runs this migrator on [entrypoint].
+  ///
+  /// This will return either a map containing only the migrated contents of
+  /// [entrypoint], or an empty map if no migration was necessary.
+  p.PathMap<String> migrateFile(String entrypoint) {
+    _patches = [];
+    var path = canonicalizePath(p.join(Directory.current.path, entrypoint));
+    var contents = File(path).readAsStringSync();
+    var syntax = Syntax.forPath(path);
+    visitStylesheet(Stylesheet.parse(contents, syntax, url: path));
+
+    var results = p.PathMap<String>();
+    if (_patches.isNotEmpty) {
+      results[path] = Patch.applyAll(_patches.first.selection.file, _patches);
+    }
+    _patches = null;
+    return results;
+  }
+
+  // Expression Tree Traversal
+
+  @override
```

Comment:
You seem to be using this inconsistently... all of the functions below are overridden, but only this is tagged. Either use it for all overrides or none.